firedrake/slate/slate.py

```diff
@@ -0,0 +1,673 @@
+    for use in symbolic linear algebra computations.
+
+    Note that this class will complain if you provide an incompatible
+    form object as input. The `Vector` class expects a form of rank 1.
+
+    :arg form: a :class:`ufl.Form` object representing a 1-form.
+    """
+
+    def __init__(self, form):
+        """Constructor for the Vector class."""
+        r = len(form.arguments())
+        if r != 1:
+            raise Exception("Cannot create a `slate.Vector` from a form with rank %d" % r)
+        self.form = form
+        self.check_integrals(form.integrals())
+        super(Vector, self).__init__()
+        # Generate integral data after tensor has been initialized
+        self.generate_integral_info(form.integrals())
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.form.arguments()
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.form.coefficients()
+
+    def __str__(self, prec=None):
+        """String representation of a SLATE Vector object."""
+        return "V_%d" % self.id
+
+    __repr__ = __str__
+
+
+class Matrix(Tensor):
+    """A matrix representation of a 2-form. This class wraps the
+    relevant information provided from a :class:`ufl.Form` object
+    for use in symbolic linear algebra computations.
+
+    Note that this class will complain if you provide an incompatible
+    form object as input. The `Matrix` class expects a form of rank 2.
+
+    :arg form: a :class:`ufl.Form` object representing a 2-form.
+    """
+
+    def __init__(self, form):
+        """Constructor for the Matrix class."""
+        r = len(form.arguments())
+        if r != 2:
+            raise Exception("Cannot create a `slate.Matrix` from a form with rank %d" % r)
+        self.form = form
+        self.check_integrals(form.integrals())
+        super(Matrix, self).__init__()
+        # Generate integral data after tensor has been initialized
+        self.generate_integral_info(form.integrals())
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.form.arguments()
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.form.coefficients()
+
+    def __str__(self, prec=None):
+        """String representation of a rank-2 tensor object in SLATE."""
+        return "M_%d" % self.id
+
+    __repr__ = __str__
+
+
+class Inverse(Tensor):
+    """An abstract SLATE class representing the tensor inverse."""
+
+    def __init__(self, A):
+        """Constructor for the Inverse class.
+
+        :arg A: a SLATE tensor."""
+        if isinstance(A, (Scalar, Vector)):
+            raise Exception("Expecting a `slate.Matrix` object, not %r" % A)
+        assert A.shape[0] == A.shape[1], "The inverse can only be computed on square tensors."
+        self.children = A
```

Comment:
Misleading name: `children` is plural, while it contains the element directly (not even a singleton list/tuple).